src/main/java/cubicchunks/worldgen/concurrency/CubeWorker.java

```diff
@@ -0,0 +1,163 @@
+package cubicchunks.worldgen.concurrency;
+
+import cubicchunks.CubicChunks;
+import cubicchunks.util.CubeCoords;
+import cubicchunks.world.cube.Cube;
+
+// TODO: Add means to properly shutdown.
+
+public abstract class CubeWorker implements Runnable {
+
+	private final String name;
+
+	private final CubeWorkerQueue queue;
+
+	private CubeCoords assignment;
+
+	// Synchronization
+
+	private final Object pauseMutex;
+
+	private boolean shouldRun;
+
+	private boolean running;
+
+	// Reporting
+
+	private long totalProcessed;
+
+	private long processedSinceReset;
+
+	private long totalDuration;
+
+	private long durationSinceReset;
+
+	// Workload management
+
+	// TODO: Make dynamic
+	private final int batchSize = 500;
+
+
+
+	public CubeWorker(String name, CubeWorkerQueue queue) {
+		this.name = name;
+		this.queue = queue;
+		this.pauseMutex = new Object();
+		this.shouldRun = false;
+		this.running = false;
+	}
+
+
+	public void assign(CubeCoords coords) {
+		synchronized (this) {
+			this.assignment = coords;
+		}
+	}
+
+	public CubeCoords getAssignment() {
+		return this.assignment;
+	}
+
+
+	public abstract boolean process(Cube cube);
+
+	public void run() {
+
+		this.pause();
+
+		while (true) {
+			this.waitUntilUnpaused();
+
+			long timeStart = System.currentTimeMillis();
+
+			int processed = 0;
+			while (this.canRun()) {
+				// Get the next cube and process it.
+				Cube cube = this.queue.poll(this);
+
+				// If there is available cube, stop processing and continue in the next tick.
+				if (cube == null) {
+					break;
+				}
+
+				// Otherwise process it.
+				if (this.process(cube)) {
+					++processed;
+				}
+			}
+
+			long timeDiff = System.currentTimeMillis() - timeStart;
+
+			// Reporting
+			this.totalDuration += timeDiff;
+			this.durationSinceReset += timeDiff;
+			this.totalProcessed += processed;
+			this.processedSinceReset += processed;
+
+			this.pause();
+		}
+	}
+
+
+	// Synchronization
+
+	public void signalPause() {
+		synchronized (this.pauseMutex) {
+			this.shouldRun = false;
```

Comment:
instead of synchronization you should make shouldRun volatile. Synchronization just means that you don't run these methods at the same time, and in this case it wouldn't be a bad thing. Volatile ensures that changes made by signalPause() in one thread are visible to canRun() in other thread.